tests/dopamine/utils/threading_utils_test.py

```diff
@@ -0,0 +1,164 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for threading_utils.py."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import mock
+import tempfile
+import threading
+
+from dopamine.agents.dqn import dqn_agent
+from dopamine.utils import threading_utils
+import tensorflow as tf
+from tensorflow import test
+
+
+def _get_internal_name(attr):
+  thread_id = str(threading.current_thread().ident)
+  return '__' + attr + '_' + thread_id
+
+
+class ThreadsTest(test.TestCase):
+  """Unit tests for threading utils."""
+
+  def testDefaultValueIsAdded(self):
+    """Tests that the default value is properly set by the helper."""
+    obj = mock.Mock()
+    threading_utils.initialize_local_attributes(obj, attr=3)
+    self.assertEqual(obj.attr_default, 3)
+
+  def testMultipleDefaultValuesAreSet(self):
+    """Tests that multiple default values are properly set by the helper."""
+    obj = mock.Mock()
+    threading_utils.initialize_local_attributes(obj, attr1=3, attr2=4)
+    self.assertEqual(obj.attr1_default, 3)
+    self.assertEqual(obj.attr2_default, 4)
+
+  def testAttributeDefaultValueIsCalled(self):
+    """Tests that getter properly uses the default value."""
+    MockClass = threading_utils.local_attributes(['attr'])(
+        type('MockClass', (object,), {}))
+    obj = MockClass()
+    obj.attr_default = 'default-value'
+    self.assertEqual(obj.attr, 'default-value')
+
+  def testDefaultValueIsRead(self):
+    """Tests that getter properly initializes the local value."""
+    MockClass = threading_utils.local_attributes(['attr'])(
+        type('MockClass', (object,), {}))
+    obj = MockClass()
+    obj.attr_default = 'default-value'
+    # Calling the attribute is expected to initialize it with the default value.
+    # Hence the pointless statement to run the getter.
+    obj.attr  # pylint: disable=pointless-statement
+    self.assertEqual(getattr(obj, _get_internal_name('attr')), 'default-value')
+
+  def testInternalAttributeIsRead(self):
+    """Tests that getter properly uses the internal value."""
+    MockClass = threading_utils.local_attributes(['attr'])(
+        type('MockClass', (object,), {}))
+    obj = MockClass()
+    setattr(obj, _get_internal_name('attr'), 'intenal-value')
+    self.assertEqual(obj.attr, 'intenal-value')
+
+  def testInternalAttributeIsSet(self):
+    """Tests that setter properly sets the internal value."""
+    MockClass = threading_utils.local_attributes(['attr'])(
+        type('MockClass', (object,), {}))
+    obj = MockClass()
+    obj.attr = 'internal-value'
```

Comment:
Here, too, I like checking that `obj.attr` can be retrieved after being set more than I like testing the internal state.  I'll stop making these comments.

Reply:
Same as other comment. I separated both types of tests.